README
pod2text CoGeX.pm > README

If this is still here it means the programmer was too lazy to create the readme file.

You can create it now by using the command shown above from this directory.

At the very least you should be able to use this set of instructions
to install the module...

perl Makefile.PL
make
make test
make install

If you are on a windows box you should use 'nmake' rather than 'make'.


Changes
Revision history for Perl module CoGeX

0.1 Mon Dec  4 13:38:26 2006
    - original version; created by ExtUtils::ModuleMaker 0.47


Makefile.PL

use ExtUtils::MakeMaker;

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME         => 'CoGeX',
    VERSION_FROM => 'lib/CoGeX.pm',                    # finds \$VERSION
    AUTHOR       => ' (elyons@nature.berkeley.edu)',
    ABSTRACT     => 'CoGeX',
    PREREQ_PM    => {
        'Benchmark'                     => 0,
        'Carp'                          => 0,
        'Class::Accessor'               => 0,
        'CoGe::Accessory::Annotation'   => 0,
        'CoGe::Accessory::genetic_code' => 0,
        'DBIx::Class'                   => 0.08,
        'DBIx::Class::Core'             => 0,
        'DBIx::Class::ResultSet'        => 0,
        'DBIx::Class::Schema'           => 0,
        'Data::Dumper'                  => 0,
        'File::Spec::Functions'         => 0,
        'LWP::Simple'                   => 0,
        'POSIX'                         => 0,
        'Test::Simple'                  => 0.44,
        'Text::Wrap'                    => 0,
        'Time::HiRes'                   => 0,
    },
);


lib/CoGeX.pm
package CoGeX;

no warnings qw(redefine);
use strict;
use warnings;


use vars qw( $VERSION );

$VERSION = 0.02;

use base 'DBIx::Class::Schema';
use base qw(Class::Accessor);
#__PACKAGE__->load_classes();
__PACKAGE__->load_namespaces();
__PACKAGE__->mk_accessors(qw(db_connection_string db_name db_passwd));

use vars qw($DEFAULT_CONNECTION_STRING $DEFAULT_NAME $DEFAULT_PASSWD);

#$DEFAULT_CONNECTION_STRING = 'dbi:mysql:dbname=coge;host=biocon.berkeley.edu;port=3306';
#$DEFAULT_CONNECTION_STRING = 'dbi:mysql:dbname=coge;host=homer.cnr.berkeley.edu;port=3306';
$DEFAULT_CONNECTION_STRING = 'dbi:mysql:dbname=coge;host=synteny.cnr.berkeley.edu;port=3306';

$DEFAULT_NAME = "coge";
$DEFAULT_PASSWD = "XXXXXXXX";

=head1 NAME

CoGeX - CoGeX

=head1 SYNOPSIS

  use CoGeX;
  This object is the API to the CoGe genomes database.  It uses DBIx::Class for managing
  relationships and access to the database.  Various other "high-level" functions are
  provided to make getting genomic data easier.


=head1 DESCRIPTION

Primary object for interacting with CoGe database system.

=head1 USAGE

  use CoGeX;

  my $s = CoGeX->connectdb(); # Biocon's ro user

  my $rs = $s->resultset('Feature')->search(
                {
                  'organism.name' => "Arabidopsis thaliana"
                },
                { join => [ 'dataset', 'organism' ] }
  );

=head1 BUGS


=head1 SUPPORT


=head1 AUTHORS

 Eric Lyons
 Brent Pedersen

=head1 COPYRIGHT

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.


=head1 SEE ALSO

perl(1).

=cut

################################################ subroutine header begin ##

=head2 dbconnect

 Usage     : use CoGeX;
             my $coge = CoGeX->dbconnect;

 Purpose   : generates the CoGe genomes API object
 Returns   : CoGeX object
 Argument  : db_connection_string=>'dbi:mysql:dbname=genomes;host=dbhostname.edu;port=3306'
             db_name=>'database_user_name'
             db_passwd=>'user_password'
             Arguments should be optional with defaults set in object so that this function can be
             called to generate db API object
 Throws    : none
 Comments  : 

See Also   : 

=cut

################################################## subroutine header end ##


sub dbconnect
  {
    my ($self, %opts) = self_or_default(@_);
    my $str = $opts{db_connection_string};
    my $name = $opts{db_name};
    my $pwd = $opts{db_passwd};
    $str = $self->db_connection_string unless $str;
    $str = $DEFAULT_CONNECTION_STRING unless $str;
    $name = $self->db_name unless $name;
    $name = $DEFAULT_NAME unless $name;
    $pwd = $self->db_passwd unless $pwd;
    $pwd = $DEFAULT_PASSWD unless $pwd;
    my $cogedb = CoGeX->connect($str, $name, $pwd );
    $cogedb->db_connection_string($str);
    $cogedb->db_name($name);
    $cogedb->db_passwd($pwd);
    return $cogedb;
  }



################################################ subroutine header begin ##

=head2 get_features_in_region

 Usage     : $object->get_features_in_region(start   => $start, 
                                             stop    => $stop, 
                                             chr     => $chr,
                                             dataset_id => $dataset->id());

 Purpose   : gets all the features in a specified genomic region
 Returns   : an array or an array_ref of feature objects (wantarray)
 Argument  : start   => genomic start position
             stop    => genomic stop position
             chr     => chromosome
             dataset_id => dataset id in database (obtained from a
                        CoGe::Dataset object)
                        of the dna seq will be returned
             OPTIONAL
             count   => flag to return only the number of features in a region
 Throws    : none
 Comments  : 

See Also   : 

=cut

################################################## subroutine header end ##

sub get_features_in_region
  {
    my $self = shift;
    my %opts = @_;
    my $start = $opts{'start'} || $opts{'START'} || $opts{begin} || $opts{BEGIN};
    $start = 1 unless $start;
    my $stop = $opts{'stop'} || $opts{STOP} || $opts{end} || $opts{END};
    $stop = $start unless defined $stop;
    my $chr = $opts{chr} || $opts{CHR} || $opts{chromosome} || $opts{CHROMOSOME};
    my $dataset_id = $opts{dataset} || $opts{dataset_id} || $opts{info_id} || $opts{INFO_ID} || $opts{data_info_id} || $opts{DATA_INFO_ID} ;
    my $count_flag = $opts{count} || $opts{COUNT};
    if ($count_flag)
      {
	return $self->resultset('Feature')->count(
						  {
						   "me.chromosome" => $chr,
						   "me.dataset_id" => $dataset_id,
						   -and=>[
							  -or=>[
								-and=>[
								       "me.stop"=>  {"<=" => $stop},
								       "me.stop"=> {">=" => $start},
								      ],
								-and=>[
								       "me.start"=>  {"<=" => $stop},
								       "me.start"=> {">=" => $start},
								      ],
								-and=>[
								       "me.start"=>  {"<=" => $start},
								       "me.stop"=> {">=" => $stop},
								      ],
							       ],
							 ],
						  },
						  {
#						   prefetch=>["locations", "feature_type"],
						  }
						 );
      }
    my @feats = $self->resultset('Feature')->search({
                 "me.chromosome" => $chr,
                 "me.dataset_id" => $dataset_id,
		 -and=>[
			-or=>[
			      -and=>[
				     "me.stop"=>  {"<=" => $stop},
				     "me.stop"=> {">=" => $start},
				    ],
			      -and=>[
				     "me.start"=>  {"<=" => $stop},
				     "me.start"=> {">=" => $start},
				    ],
			      -and=>[
				     "me.start"=>  {"<=" => $start},
				     "me.stop"=> {">=" => $stop},
				    ],
			     ],
		       ],
						    },
						    {
						     prefetch=>["locations", "feature_type"],
						     order_by=>"me.start",
						    }
						   );
    return wantarray ? @feats : \@feats;
  }

sub get_features_in_region_split
  {
    my $self = shift;
    my %opts = @_;
    my $start = $opts{'start'} || $opts{'START'} || $opts{begin} || $opts{BEGIN};
    $start = 0 unless $start;
    my $stop = $opts{'stop'} || $opts{STOP} || $opts{end} || $opts{END};
    $stop = $start unless defined $stop;
    my $chr = $opts{chr} || $opts{CHR} || $opts{chromosome} || $opts{CHROMOSOME};
    my $dataset_id = $opts{dataset} || $opts{dataset_id} || $opts{info_id} || $opts{INFO_ID} || $opts{data_info_id} || $opts{DATA_INFO_ID} ;

    my @startfeats = $self->resultset('Feature')->search({
                 "me.chromosome" => $chr,
                 "me.dataset_id" => $dataset_id,
                 -and => [
                   "me.stop"=> {">=" => $start},
                   "me.stop"=>  {"<=" => $stop},
                 ],
						     },
                 {
                   prefetch=>["locations", "feature_type"],
                 }
						   );
    my @stopfeats = $self->resultset('Feature')->search({
                 "me.chromosome" => $chr,
                 "me.dataset_id" => $dataset_id,
                 -and => [
                   "me.start"=>  {">=" => $start},
                   "me.start"=> {"<=" => $stop},
                 ],
						     },
                 {
                   prefetch=>["locations", "feature_type"],
                 }
						   );

    my %seen;
    my @feats;

    foreach my $f ( @startfeats ) {
      if ( not exists $seen{ $f->id() } ) {
        $seen{$f->id()}+=1;
        push( @feats, $f );
      }
    }

    foreach my $f ( @stopfeats ) {
      if ( not exists $seen{ $f->id() } ) {
        $seen{$f->id()}+=1;
        push( @feats, $f );
      }
    }

    return wantarray ? @feats : \@feats;
  }

################################################ subroutine header begin ##

=head2 count_features_in_region

 Usage     : $object->count_features_in_region(start   => $start, 
                                             stop    => $stop, 
                                             chr     => $chr,
                                             dataset_id => $dataset->id());

 Purpose   : counts the features in a specified genomic region
 Returns   : an integer
 Argument  : start   => genomic start position
             stop    => genomic stop position
             chr     => chromosome
             dataset_id => dataset id in database (obtained from a
                        CoGe::Dataset object)
                        of the dna seq will be returned
 Throws    : none
 Comments  : 

See Also   : 

=cut

################################################## subroutine header end ##

sub count_features_in_region
  {
    my $self = shift;
    my %opts = @_;
    return $self->get_features_in_region (%opts, count=>1);
  }

sub get_current_datasets_for_org
  {
    my $self = shift;
    warn 'THIS METHOD (get_current_datasets_for_org) IS OBSELETE.  PLEASE CALL $org_obj->current_datasets()\n';
    my %opts = @_ if @_ >1;
    my $orgid = $opts{org} || $opts{orgid} || $opts{organism};
    $orgid = shift unless $orgid;
    return unless $orgid;
    my ($org) = $self->resultset('Organism')->resolve($orgid);
    return unless $org;
    return $org->current_datasets();
  }

sub log_user
  {
    my $self = shift;
    my %opts = @_;
    my $user = $opts{user};
    my $session = $opts{session};
    my $uid = ref($user) =~ /User/ ? $user->id : $user;
    unless ($uid =~ /^\d+$/)
      {
	warn "Error adding user_id to User_session.  Not a valid uid: $uid\n";
	return;
      }
    #FIRST REMOVE ALL ENTRIES FOR THIS USER
    foreach my $item ($self->resultset('UserSession')->search(session=>$session))
      {
	next unless $item;
	$item->delete;
      }
    #ADD NEW ENTRY
    my $item = $self->resultset('UserSession')->create({user_id=>$uid, date=>\'NOW()', session=>$session});
    return $item;
  }

sub self_or_default 
    { #adapted from CGI.pm
      shift @_ if $_[0] eq "CoGeX";
      my $Q;
      unless (
	      defined($_[0]) && 
	      (ref($_[0]) eq 'CoGeX')# || UNIVERSAL::isa($_[0],'CoGeX')) # slightly optimized for common case
	     ) 
	{
	  $Q = CoGeX->new unless defined($Q);
	  unshift(@_,$Q);
	}
      return wantarray ? @_ : $Q;
    }


1;


lib/CoGeX/Result/FeatureSummary.pm
package CoGeX::Result::FeatureSummary;

use strict;
use warnings;

use base 'DBIx::Class::Core';

=head1 NAME

CoGeX::Result::FeatureSummary

=head1 SYNOPSIS

This object defines a view that can be used to obtain summarized feature
information for a dataset in the CoGe database.

=head1 DESCRIPTION

Has columns:
C<feature_type_id>
Type: INT, Default: undef, Nullable: no, Size: 11

C<feature_type_name>
Type: VARCHAR, Default: "", Nullable: no, Size 255

C<feature_count>
Type: INT, Default: undef, Nullable: no, Size: 11

=head1 USAGE

  use CoGeX;

=head1 METHODS

=cut

__PACKAGE__->table_class('DBIx::Class::ResultSource::View');

__PACKAGE__->table("feature");
__PACKAGE__->add_columns(
    "feature_type_id",
    {   data_type     => "INT",
        default_value => undef,
        is_nullable   => 0,
        size          => 11
    },
    "feature_type_name",
    {   data_type     => "VARCHAR",
        default_value => "",
        is_nullable   => 0,
        size          => 255
    },
    "feature_count",
    {   data_type     => "INT",
        default_value => undef,
        is_nullable   => 0,
        size          => 11
    },
);

__PACKAGE__->result_source_instance->is_virtual(1);

__PACKAGE__->result_source_instance->view_definition(<<'END_OF_DEFINITION');
    SELECT
        f.feature_type_id   AS feature_type_id,
        t.name              AS feature_type_name,
        COUNT(f.feature_id) AS feature_count
    FROM feature f JOIN feature_type t ON t.feature_type_id = f.feature_type_id
    WHERE f.dataset_id = ?
    GROUP BY f.feature_type_id
END_OF_DEFINITION

1;

=head1 BUGS


=head1 SUPPORT


=head1 AUTHORS

 Eric Lyons
 Brent Pedersen

=head1 COPYRIGHT

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.


=head1 SEE ALSO

=cut


scripts/dump_GenomicSequence_data.pl
#!/usr/bin/perl

use strict;
use Data::Dumper;
use CoGeX;
use File::Path;
use File::Spec::Functions;
use POSIX;

my $BASEDIR = "/opt/apache/CoGe/data/genomic_sequence";
my $connstr = 'dbi:mysql:dbname=genomes_new;host=biocon.berkeley.edu;port=3306';
my $coge = CoGeX->connect($connstr, 'coge', 'XXXXXXXX' );

foreach my $item ($coge->resultset('Dataset')->all())
  {
    next unless $item->sequence_type;
    my $dsid = "dataset_".$item->id;
    my $type_id = "seqtype_".$item->sequence_type->id;
    my $chr_total = $item->chromosomes;
    $chr_total = scalar @$chr_total;
    my $orgid = "org_".$item->organism->id;
    my $dir = catfile($BASEDIR,ceil($item->organism->id/1000),$orgid,$type_id,$dsid);
    my $chr_level =1;
    mkpath($dir."/".$chr_level);
    my $chr_count = 0;
    foreach my $chr(sort $item->chromosomes)
      {
	my $seq = $item->genomic_sequence(chr=>$chr, start=>1, stop=>$item->last_chromosome_position($chr), skip_length_check=>1);
	open (OUT, ">$dir/$chr_level/$chr");
	print OUT $seq;
	close OUT;
	$chr_count++;
	unless ($chr_count%1000)
	  {
	    $chr_level++;
	    mkpath($dir."/".$chr_level);
	  }
      }
#    print $dir."/".$chr_level,"\t",$chr_count, "\t", $chr_total,"\n" if $chr_total > 1000;
  }


lib/CoGeX/Result/GenomeFeatureSummary.pm
package CoGeX::Result::GenomeFeatureSummary;

use strict;
use warnings;

use base 'DBIx::Class::Core';

=head1 NAME

CoGeX::Result::FeatureSummary

=head1 SYNOPSIS

This object defines a view that can be used to obtain summarized feature
information for a dataset group in the CoGe database.

=head1 DESCRIPTION

Has columns:
C<feature_type_id>
Type: INT, Default: undef, Nullable: no, Size: 11

C<feature_type_name>
Type: VARCHAR, Default: "", Nullable: no, Size: 255

C<feature_count>
Type: INT, Default: undef, Nullable: no, Size: 11

=head1 USAGE

  use CoGeX;

=head1 METHODS

=cut

__PACKAGE__->table_class('DBIx::Class::ResultSource::View');

__PACKAGE__->table("feature");
__PACKAGE__->add_columns(
    "feature_type_id",
    {   data_type     => "INT",
        default_value => undef,
        is_nullable   => 0,
        size          => 11
    },
    "feature_type_name",
    {   data_type     => "VARCHAR",
        default_value => "",
        is_nullable   => 0,
        size          => 255
    },
    "feature_count",
    {   data_type     => "INT",
        default_value => undef,
        is_nullable   => 0,
        size          => 11
    },
);

__PACKAGE__->result_source_instance->is_virtual(1);

__PACKAGE__->result_source_instance->view_definition(<<'END_OF_DEFINITION');
    SELECT
        f.feature_type_id   AS feature_type_id,
        t.name              AS feature_type_name,
        COUNT(f.feature_id) AS feature_count
    FROM feature f
    JOIN feature_type t      ON t.feature_type_id = f.feature_type_id
    JOIN dataset_connector c ON c.dataset_id = f.dataset_id
    WHERE c.dataset_group_id = ?
    GROUP BY f.feature_type_id
END_OF_DEFINITION

1;

=head1 BUGS


=head1 SUPPORT


=head1 AUTHORS

 Eric Lyons
 Brent Pedersen

=head1 COPYRIGHT

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.


=head1 SEE ALSO

=cut


lib/CoGeX/Result/Organism.pm
package CoGeX::Result::Organism;

# Created by DBIx::Class::Schema::Loader v0.03009 @ 2006-12-01 18:13:38

use strict;
use warnings;
use CoGeX::ResultSet::Organism;
use base 'DBIx::Class::Core';

=head1 NAME

CoGeX::Organism

=head1 SYNOPSIS

This object uses the DBIx::Class to define an interface to the C<organism> table in the CoGe database.
The C<organism> table contains the name, description, and normalized name for an organism.

=head1 DESCRIPTION

Has columns:
C<organism_id> (Primary Key)
Type: INT, Default: undef, Nullable: no, Size: 11
Primary identification key for table.

C<name>
Type: VARCHAR, Default: "", Nullable: no, Size: 200
Organism name.

C<description>
Type: VARCHAR, Default: undef, Nullable: yes, Size: 255
Organism description.

C<normalized_name>
Type:VARCHAR, Default: "", Nullable: no, Size: 200
File-system 'safe' (only alphanumeric characters and underscores) version of 'name' field above.


Relates to CCoGeX::Result::DatasetGroup> via C<organism_id>, in a one-to-many relationship.

=head1 USAGE

  use CoGeX;

=head1 METHODS

=cut

__PACKAGE__->table("organism");
__PACKAGE__->add_columns(
  "organism_id",
  { data_type => "INT", default_value => undef, is_nullable => 0, size => 11 },
  "name",
  { data_type => "VARCHAR", default_value => "", is_nullable => 0, size => 200 },
  "description",
  {
    data_type => "VARCHAR",
    default_value => undef,
    is_nullable => 1,
    size => 255,
  },
  "restricted",  { data_type => "int", default_value => "0", is_nullable => 0, size => 1 },
  "access_count",  { data_type => "int", default_value => "0", is_nullable => 0, size => 10 },
);
__PACKAGE__->set_primary_key("organism_id");

__PACKAGE__->has_many("dataset_groups" => "CoGeX::Result::DatasetGroup", 'organism_id');


################################################ subroutine header begin ##

=head2 current_genome

 Usage     : 
 Purpose   : 
 Returns   : 
 Argument  : 
 Throws    : 
 Comments  : 
           : 

See Also   : 

=cut

################################################## subroutine header end ##

sub current_genome
  {
    my $self = shift;
    my %opts = @_;
    my $gstid = $opts{type} || $opts{genomic_sequence_type} || $opts{sequence_type} || $opts{gstid};
    $gstid = 1 unless $gstid;
    $gstid = ref($gstid) =~/Type/ ? $gstid->id : $gstid;
    my ($dsg) = $self->dataset_groups({genomic_sequence_type_id=>$gstid},
				      {'order_by' => 'me.version desc',
				       join=>[{'dataset_connectors'=>'dataset'}, 'organism'],
				      prefetch=>[{'dataset_connectors'=>'dataset'}, 'organism']
				      },
				     );
    return $dsg;
  }


################################################ subroutine header begin ##

=head2 current_datasets

 Usage     : 
 Purpose   : 
 Returns   : 
 Argument  : 
 Throws    : 
 Comments  : 
           : 

See Also   : 

=cut

################################################## subroutine header end ##

sub current_datasets
  {
    my $self = shift;
    my %opts = @_;
    my $dgs = $self->current_genome(%opts);
    return $dgs->datasets();
  }


################################################ subroutine header begin ##

=head2 current_datasets_old

 Usage     : 
 Purpose   : 
 Returns   : 
 Argument  : 
 Throws    : 
 Comments  : 
           : 

See Also   : 

=cut

################################################## subroutine header end ##

sub current_datasets_old
  {
    my $self = shift;
    my %opts = @_;
    my $type = $opts{type} || $opts{genomic_sequence_type} || $opts{sequence_type} || $opts{gstid};
    $type =1 unless $type;
    my %data;
    my $typeid;
    $typeid = ref($type) =~/Type/ ? $type->id : $type;
    my $version;
    my $contig_set =0;
    ds_loop: foreach my $ds ($self->datasets({},{distict=>'version',order_by=>'version desc'}))
      {
	next unless $ds->sequence_type && $ds->sequence_type->id eq $typeid;
	$version = $ds->version unless $version;
	$version = $ds->version if $ds->version > $version;
#	next unless $version == $ds->version;
	my @chrs = $ds->get_chromosomes;
	$contig_set = 1 if scalar @chrs > 50; #more than 50 chromosome?  probably a contig dataset
	foreach my $chr (@chrs)
	  {
	    #this is a hack but the general problem is that some organisms have different chromosomes at different versions, however, partially complete genomes will have many contigs and different versions will have different contigs.  So, to get around this, there is a check to see if the chromosome name has contig in it, if so, then only the most current version is used.  Otherwise, all versions are game.
	    next unless $chr;
	    if ($chr =~ /contig/i || $chr=~/scaffold/i || $contig_set)
	      {
		$contig_set = 1;
		next ds_loop if $ds->version ne $version;
	      }
	    $data{$chr} = $ds unless $data{$chr};# || $contig_set;
	    $data{$chr} = $ds if $ds->version > $data{$chr}->version;
	  }
      }
    %data = map {$_->id,$_} values %data;
    return wantarray ? values %data : [values %data];
  }


################################################ subroutine header begin ##

=head2 genomic_sequence_types

 Usage     : 
 Purpose   : 
 Returns   : 
 Argument  : 
 Throws    : 
 Comments  : 

See Also   : 

=cut

################################################## subroutine header end ##

sub genomic_sequence_types
  {
    my $self = shift;
    my %data;
    foreach my $ds ($self->datasets)
      {
        my $types_ref = $ds->sequence_type;
        if (defined $types_ref)
          {
            if ( ref $types_ref ne 'ARRAY' )
              {
                $types_ref = [ $types_ref ];
              }
            foreach my $type (@{$types_ref})
              {
                $data{$type->id} = $type;
              }
          }
      }
    return wantarray ? values %data : [values %data];
  }


################################################ subroutine header begin ##

=head2 types

 Usage     : 
 Purpose   : See genomic_sequence_types()
 Returns   : 
 Argument  : 
 Throws    : 
 Comments  : Alias for genomic_sequence_types() method.

See Also   : genomic_sequence_types()

=cut

################################################## subroutine header end ##

sub types
  {
    return shift->genomic_sequence_types(@_);
  }


################################################ subroutine header begin ##

=head2 datasets

 Usage     : 
 Purpose   : 
 Returns   : 
 Argument  : 
 Throws    : 
 Comments  : 

See Also   : 

=cut

################################################## subroutine header end ##

sub datasets
  {
    my $self = shift;
    my %opts = @_;
    my %ds;
    map {$ds{$_->id}=$_} map{$_->datasets} $self->dataset_groups;
    return wantarray ? values %ds : [values %ds];
  }

1;



=head1 BUGS


=head1 SUPPORT


=head1 AUTHORS

 Eric Lyons
 Brent Pedersen

=head1 COPYRIGHT

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.


=head1 SEE ALSO

=cut
